File: Julia/smolyak.jl
using Iterators
using Cartesian
# import PyPlot
reload("util")
import Base.show

## ------------- ##
#- Generic Tools -#
## ------------- ##

counting_coef(d, mu, i) = (-1) ^ (d + mu - i) * choose(d - 1, d + mu - i)


function num_grid_pts(d::Int, mu::Int)

    if mu == 1
        return int(2d - 1)
    elseif mu == 2
        return int(1 + 4d + 4d*(d-1)/2)
    elseif mu == 3
        return int(1 + 8d + 12d*(d-1)/2 + 8d*(d-1)*(d-2)/6)
    else
        error("We only know the number of grid points for mu \\in [1, 2, 3]")
    end
end


function m_i(i::Int)
    if i < 0
        error("DomainError: i must be positive")
    elseif i == 0
        return 0
    elseif i == 1
        return 1
    else
        return 2^(i - 1) + 1
    end
end


function cheby2n{T<:Number}(x::Array{T, 1}, n::Int; kind::Real=1)
    # Evaluates the first n+1 Chebyshev polynomials of the 'kind' kind at x
    # NOTE: This will only work when kind = 1 or kind = 2
    # NOTE: We evaluate the first n+1, with the first dimension being all 1's

    dim = length(x)
    results = Array(T, dim, n + 1)
    results[:, 1] = 1.
    results[:, 2] = kind * x
    for i=3:n+1
        results[:, i] = 2x .* results[:, i - 1] - results[:, i - 2]
    end
    return results
end


function cheby2n{T<:Number}(x::Array{T, 2}, n::Int; kind::Real=1)
    # Evaluates the first n+1 Chebyshev polynomials of the 'kind' kind at x
    # NOTE: This will only work when kind = 1 or kind = 2
    # NOTE: We evaluate the first n+1, with the first dimension being all 1's

    dim = size(x)
    results = Array(T, dim..., n + 1)
    results[:, :, 1] = 1.
    results[:, :, 2] = kind * x
    for i=3:n+1
        results[:, :, i] = 2x .* results[:, :, i - 1] - results[:, :, i - 2]
    end
    return results
end


function S_n(n::Int)
    # Compute the set S_n. All points are extrema of Chebyshev polynomials

    if n < 1
        error("DomainError: n must be positive")
    end

    if n == 1
        return [0.]
    end

    m = m_i(n)
    j = [1:m]
    pts = cos(pi * (j - 1.) / (m - 1))::Array{Float64, 1}
    for i = 1:size(pts, 1)
        pts[i] = abs(pts[i]) < 1e-12 ? 0.0 : pts[i]
    end
    return pts
end


function A_n(n::Int)
    if n < 1
        error("DomainError: n must be positive")
    end

    if n == 1
        return [0.]
    elseif n == 2
        return [-1., 1.]
    end

    sn = S_n(n)
    return sn[[2:2:size(sn, 1)]]
end


function A_chain(n::Int)
    # Return Dict of all A_n from n to 1. Keys are n, values are arrays

    sn = S_n(n)

    a = Dict{Int, Array{Float64, 1}}()
    sizehint(a, n)

    # These are constant and don't follow the pattern.
    a[1] = [0.]
    a[2] = [-1., 1.]

    for i=n:-1:3
        # push!(a, sn[[2:2:size(sn, 1)]])
        a[i] = sn[[2:2:size(sn, 1)]]
        sn = sn[[1:2:size(sn, 1)]]
    end

    return a
end


function cheb_dict(mu::Int)
    pts = S_n(mu + 1)
    max_poly = length(pts)

    # TODO: Work on figuring out the types here
    # cd = Dict{Int, Dict{Float64, Float64}}()
    cd = Dict()

    for phi_n=1:max_poly
        cd[phi_n] = {pt => t_pt for (pt, t_pt) in
                     zip(pts, cheby_eval(pts, phi_n))}
    end

    return cd
end


function cheby_eval(x::Array{Float64}, n::Int)
    pv = ones(size(x)...)
    cv = x

    if n == 1
        return pv
    elseif n == 2
        return cv
    else
        for i=3:n
            tmp = 2 .* x .* cv - pv
            pv = cv
            cv = tmp
        end
        return cv
    end
end


## --------------- ##
#- Isotropic Grids -#
## --------------- ##


function smol_inds(d::Int, mu::Int)
    p_vals = [1:mu + 1]

    poss_inds = cell(1)  # see below

    for el in Task(()-> comb_with_replacement(p_vals, d))
        if d < sum(el) <= d + mu
            push!(poss_inds, el)
        end
    end

    # cell(1) gives undefined reference in first slot. We remove it here
    poss_inds = poss_inds[2:]

    true_inds = Any[ones(Int64, d)]
    for val in poss_inds
        for el in Task(()->pmute(val))
            push!(true_inds, el)
        end
    end

    return true_inds
end


function base_inds(n::Int)
    max_ind = m_i(n)
    phi = Dict{Int, Array{Int64, 1}}()
    phi[1] = [1]
    phi[2] = [2, 3]
    low_ind = 4  # current lower index
    for i = 3:n
        high_ind = m_i(i)
        phi[i] = [low_ind:high_ind]
        low_ind = high_ind + 1
    end

    return phi
end


function sparse_grid(d::Int, mu::Int)
    # Use disjoint Smolyak sets to construct Smolyak grid

    true_inds = smol_inds(d, mu)
    An = A_chain(mu + 1)
    return vcat([ cartprod([An[i] for i in el]) for el in true_inds]...)
end


function grid_base(d::Int, mu::Int)
    # Build B matrix for interpolating polynomial

    true_inds = smol_inds(d, mu)
    An = base_inds(mu + 1)
    return vcat([ cartprod([An[i] for i in el]) for el in true_inds]...)
end

### type SGrid

type SGrid
    d::Int
    mu::Int
    grid::Array{Float64, 2}
    inds::Array{Any, 1}

    function SGrid(d::Int, mu::Int, grid::Array{Float64, 2}, inds::Array{Any, 1})
        if d < 2
            error("You passed d = $d. d must be greater than 1")
        end
        if mu < 1
            error("You passed mu = $mu. mu must be greater than 1")
        end
        new(d, mu, grid, inds)
    end
end


# Add convenience constructor to just pass d, mu
SGrid(d::Int, mu::Int) = SGrid(d, mu, sparse_grid(d, mu), smol_inds(d, mu))


function show(io::IO, sg::SGrid)
    npoints = size(sg.grid, 1)
    msg = "Smolyak Grid:\n\td: $(sg.d)\n\tmu: $(sg.mu)\n\tnpoints: $npoints"
    print(io, msg)
end


### Build B-matrix for grid object


function Bmat(sg::SGrid)
    # Compute the matrix B from equation 22 in JMMV 2013
    # Naive translation of dolo.numeric.interpolation.smolyak.SmolyakBasic

    # TODO: check to see if I should swap the k and : in reduce. It might be
    #       more scalable if I do. I would need pass sg.grid' to cheby2n

    Ts = cheby2n(sg.grid, m_i(sg.mu + 1))
    n = size(sg.grid, 1)
    d = sg.d
    b_inds = grid_base(sg.d, sg.mu)
    B = Array(Float64, n, n)
    for ind = 1:n
        B[:, ind] = reduce(.*, {slice(Ts, :, k, b_inds[ind, k]) for k=1:d})
    end
    return B
end


## ----------------- ##
#- Anisotropic Grids -#
## ----------------- ##


function smol_inds(d::Int, mu::Array{Int, 1})
    # Compute indices needed for anisotropic smolyak grid given number of
    # dimensions d and a vector of mu parameters mu

    if length(mu) != d
        error("ValueError: mu must have d elements.")
    end

    mu_max = maximum(mu)
    mup1 = mu + 1

    p_vals = [1:mu_max + 1]

    poss_inds = cell(1)  # see below

    for el in Task(()-> comb_with_replacement(p_vals, d))
        if d < sum(el) <= d + mu_max
            push!(poss_inds, el)
        end
    end

    # cell(1) gives undefined reference in first slot. We remove it here
    poss_inds = poss_inds[2:]

    true_inds = Any[ones(Int64, d)]
    for val in poss_inds
        for el in Task(()->pmute(val))
            if all(el .<= mup1)
                push!(true_inds, el)
            end
        end
    end

    return true_inds
end



# function plot(sg::SGrid)
#     g = sg.grid
#     if size(g, 2) == 2
#         PyPlot.plot(g[:, 1], g[:, 2], "b.")
#         PyPlot.xlim((-1.5, 1.5))
#         PyPlot.ylim((-1.5, 1.5))
#         PyPlot.title("Smolyak grid: \$d=$(sg.d), \\; \\mu=$(sg.mu)\$")
#     elseif size(g, 2) == 3
#         PyPlot.scatter3D(g[:, 1], g[:, 2], g[:, 3], "b.")
#         PyPlot.title("Smolyak grid: \$d=$(sg.d), \\; \\mu=$(sg.mu)\$")
#     else
#         error("ERROR: can only plot 2d or 3d grids")
#     end
# end

#----------
# Test case from dolo
#----------

# fun(x, y) = exp(- x .^ 2 - y .^ 2)

File: Python/smolyak.py
"""
This file contains a class that builds a Smolyak Grid.  The hope is that
it will eventually contain the interpolation routines necessary so that
the given some data, this class can build a grid and use the Chebyshev
polynomials to interpolate and approximate the data.

Method based on Judd, Maliar, Maliar, Valero 2013 (W.P)

Authors: Chase Coleman and Spencer Lyon

"""
import sys
from operator import mul
from itertools import product, combinations_with_replacement
from itertools import chain
import numpy as np
from scipy.linalg import lu
import matplotlib.pyplot as plt
import pandas as pd
from mpl_toolkits.mplot3d import Axes3D
from smolyak_utils import *


def a_chain(n):
    """
    This method finds all of the unidimensional disjoint sets
    that we will use to construct the grid.  It improves on
    past algorithms by noting that A_{n} = S_{n}[evens] except for
    A_1 = {0} and A_2 = {-1, 1}. Additionally, A_{n} = A_{n+1}[odds]
    This prevents the calculation of these nodes repeatedly.  Thus
    we only need to calculate biggest of the S_n's to build the
    sequence of A_n's

    """

    # # Start w finding the biggest Sn(We will subsequently reduce it)
    Sn = _s_n(n)

    A_chain = {}
    A_chain[1] = [0.]
    A_chain[2] = [-1., 1.]

    # Need a for loop to extract remaining elements
    for seq in xrange(n, 2, -1):
        num = Sn.size
        # Need odd indices in python because indexing starts at 0
        A_chain[seq] = tuple(Sn[range(1, num, 2)])
        # A_chain.append(list(Sn[range(1, num, 2)]))
        Sn = Sn[range(0, num, 2)]

    return A_chain

def _s_n(n):
    """
    This method finds the element S_n for the Chebyshev Extrema
    """

    if n == 1:
        return np.array([0.])

    # Apply the necessary transformation to get the nested sequence
    m_i = 2**(n-1) + 1

    # Create an array of values that will be passed in to calculate
    # the set of values
    comp_vals = np.arange(1., m_i + 1.)

    # Values are - cos(pi(j-1)/(n-1)) for j in [1, 2, ..., n]
    vals = -1. * np.cos(np.pi*(comp_vals - 1.)/(m_i-1.))
    vals[np.where(np.abs(vals) < 1e-14)] = 0.0

    return vals

def smol_inds(d, mu):
    """
    This method finds all of the indices that satisfy the requirement
    that d \leq \sum_{i=1}^d \leq d + \mu.  Once we have these, then
    they can be used to build both the grid and the polynomial
    basis.

    Notes
    =====
    This method sets the attribute smol_inds
    """
    # Need to capture up to value mu + 1 so in python need mu+2
    possible_values = range(1, mu + 2)

    # find all (i1, i2, ... id) such that their sum is in range
    # we want; this will cut down on later iterations
    poss_inds = [el for el in combinations_with_replacement(possible_values, d)
                  if d < sum(el) <= d+mu]

    true_inds = [[el for el in permute(list(val))] for val in poss_inds]

    # Add the d dimension 1 array so that we don't repeat it a bunch
    # of times
    true_inds.extend([[[1]*d]])

    tinds = list(chain.from_iterable(true_inds))

    return tinds

def build_grid(d, mu):
    """
    This method builds a grid for the object

    Notes
    =====
    This method sets the attribute grid
    """
    # Get An chain
    An = a_chain(mu + 1)

    points = []

    # Need to get the correct indices

    tinds = smol_inds(d, mu)

    for el in tinds:
        temp = [An[i] for i in el]
        # Save these indices that we iterate through because
        # we need them for the chebyshev polynomial combination
        # inds.append(el)
        points.extend(list(product(*temp)))

    grid = pd.lib.to_object_array_tuples(points).astype(float)

    return grid

def phi_chain(n):
    """
    Finds the disjoint sets of aphi's that will be used to compute
    which functions we need to calculate
    """

    # First create a dictionary
    aphi_chain = {}

    aphi_chain[1] = [1]
    aphi_chain[2] = [2, 3]

    curr_val = 4
    for i in xrange(3, n+1):
        end_val = 2**(i-1) + 1
        temp = range(curr_val, end_val+1)
        aphi_chain[i] = temp
        curr_val = end_val+1

    return aphi_chain

def poly_inds(d, mu):
    """
    This function builds the indices of the basis polynomials that
    will be used to interpolate.
    """
    inds = smol_inds(d, mu)
    aphi = phi_chain(mu + 1)

    # Bring in polynomials
    # cheb_dict = self.calc_chebvals()

    base_polys = []

    for el in inds:
        temp = [aphi[i] for i in el]
        # Save these indices that we iterate through because
        # we need them for the chebyshev polynomial combination
        # inds.append(el)
        base_polys.extend(list(product(*temp)))

    return base_polys

def build_B(d, mu, grid):
    """
    This function builds the matrix B that will be used to calc
    the interpolation coefficients for a given set of data.

    Notes
    =====
    This method sets the attributes B, B_L, B_U
    """
    Ts = cheby2n(grid.T, m_i(mu + 1))
    base_polys = poly_inds(d, mu)
    n = len(grid)
    B = np.empty((n, n), order='F')
    for ind, comb in enumerate(base_polys):
        B[:, ind] = reduce(mul, [Ts[comb[i] - 1, i, :]
                           for i in range(d)])

    return B

def cheby_eval(x, n):
    past_val = np.ones_like(x)
    curr_val = x

    if n == 1:
        return past_val

    if n == 2:
        return curr_val

    for i in xrange(3, n + 1):
        temp = 2*x*curr_val - past_val
        past_val = curr_val
        curr_val = temp

    return curr_val

def cheby2n(x, n):
    # computes the chebychev polynomials of the first kind
    dim = x.shape
    results = np.zeros((n+1, ) + dim)
    results[0, ...] = np.ones(dim)
    results[1, ...] = x
    for i in range(2,n+1):
        results[i,...] = 2 * x * results[i-1,...] - results[i-2,...]
    return results


def num_grid_points(d, mu):
    if mu == 1:
        return 2*d + 1

    if mu == 2:
        return 1 + 4*d + 4*d*(d-1)/2.

    if mu == 3:
        return 1 + 8*d + 12*d*(d-1)/2. + 8*d*(d-1)*(d-2)/6.

def m_i(i):
    if i < 0:
        raise ValueError('i must be positive')
    elif i == 0:
        return 0
    elif i == 1:
        return 1
    else:
        return 2**(i - 1) + 1


class SmolyakGrid(object):
    """
    This class currently takes a dimension and a degree of polynomial
    and builds the Smolyak Sparse grid.  We base this on the work by
    Judd Maliar Maliar and Valero.  Hope to obtain speed ups beyond
    what they achieved.

    Attributes:
    ------------------
    d : scalar : integer
        This is the dimension of grid that you are building

    mu : scalar : integer
        mu is a parameter that defines the fineness of grid that we
        want to build


    Methods:
    --------

    a_chain : This builds the disjoint sets of basis points

    _s_n : This builds the joint set of all basis points

    _smol_inds : This constructs the indices that satisfy the
                      constraint d <= |i| <= d + mu

    build_grid : This method builds the sparse grid

    phi_chain : Builds the disjoint sets of basis polynomials subscripts
                 (1, 2) = \phi_1 \phi_2 etc...

    poly_inds : Combines the basis polynomials in a similar
                              fashion as build_grid wrt to points

    build_B : Builds the B matrix that will be used to interpolate

    plot_grid : Pretty obvious the function of this method... Plots grid

    Attributes:
    -----------

    d::Int  # number of dimensions
    mu::Int  # density parameter
    grid::Matrix{Float64}  # Smolyak grid
    inds::Array{Any, 1}  # Smolyak indices
    B::Matrix{Float64}  # matrix representing interpoland
    B_L::Matrix{Float64}  # L from LU decomposition of B
    B_U::Matrix{Float64}  # U from LU decomposition of B

    """

    def __init__(self, d, mu, do="all"):
        """
        Parameters
        ----------
        d : scalar : integer
            This is the dimension of grid that you are building

        mu : scalar : integer
            mu is a parameter that defines the fineness of grid that we
            want to build

        do : string : string
            do specifies whether you just want to build the grid or
            whether it should build the whole object.  Only takes values
            of "all" or "grid".  Default is "all"
        """
        self.d = d
        self.mu = mu

        if mu < 1:
            raise ValueError('The parameter mu needs to be > 1.')

        if d <= 1:
            raise ValueError('You are trying to build a one dimensional\
                             grid.')

        self.grid = build_grid(self.d, self.mu)
        self.inds = smol_inds(self.d, self.mu)
        if do == "all":
            self.B = build_B(self.d, self.mu, self.grid)

        # Compute LU decomposition
        l, u = lu(self.B, True)  # pass permute_l as true. See scipy docs
        self.B_L = l
        self.B_U = u



    def __repr__(self):
        msg = "Smolyak Grid:\n\td: {0} \n\tmu: {1} \n\tnpoints: {2}"
        return msg.format(self.d, self.mu, self.grid.shape[0])

    def __str__(self):
        return str(self.__repr__)

    def plot_grid(self):
        grid = self.grid
        if grid.shape[1] == 2:
            xs = grid[:, 0]
            ys = grid[:, 1]
            fig = plt.figure()
            ax = fig.add_subplot(111)
            ax.scatter(xs, ys)
            ax.grid(True, linestyle='--', color='0.75')
            plt.show()
        elif grid.shape[1] == 3:
            xs = grid[:, 0]
            ys = grid[:, 1]
            zs = grid[:, 2]
            fig = plt.figure()
            ax = fig.add_subplot(111, projection='3d')
            ax.scatter(xs, ys, zs)
            ax.grid(True, linestyle='--', color='0.75')
            plt.show()
        else:
            raise ValueError('Can only plot 2 or 3 dimensional problems')
